keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1392 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today, the scheduler surfaces a specific pod condition: `Scheduled` that clearly
+identifies whether a pod got scheduled by the scheduler and when scheduling
+completed. However, no specific conditions around initialization of successfully
+scheduled pods from the perspective of starting and completion of pod sandbox
+creation is surfaced to cluster administrators in a scoped and consumable
+fashion.
+
+Dedicated conditions around starting and completion of pod sandbox creation will
+benefit cluster operators (especially of multi-tenant clusters) who are
+responsible for configuration and operational aspects of the various components
+responsible for pod sandbox creation (CSI plugins, CRI runtime and associated
+runtime handlers, CNI plugins, etc).  Along with the conditions themselves,
+measuring the transitions of these conditions enable cluster administrators
+to surface aggregate latency of all the components involved in pod
+initialization as an SLI and publish SLOs to those who launch workloads
+on the cluster.
+
+Custom pod controllers/operators can use dedicated conditions around starting
+and completion of pod sandbox creation to make better decisions around how to
+reconcile a pod failing to become ready. As a specific example, a custom
+controller for managing pods that refer to PVCs associated with node local
+storage (e.g. Rook-Ceph) may decide to recreate PVCs (based on a specified PVC
+template) if the sandbox creation is repeatedly failing to complete. Such a
+controller can leave PVCs intact and only recreate pods if sandbox creation
+completes successfully but the pod's containers fail to become ready.
+
+There is an existing pod condition: `Initialized` that tracks execution of init
+containers. For pods without init containers, the `Initialized` condition is
+set when the Kubelet starts to process a pod before any sandbox creation
+activities start. For pods with init containers, the `Initialized` condition
+is set when init containers have been pulled and executed to completion.
+Therefore, the existing `Initialized` condition is insufficient and inaccurate
+for tracking starting and completion of sandbox creation of all pods in a
+cluster. This distinction becomes especially relevant in multi-tenant clusters
+where individual tenants own the pod specs (including the set of init
+containers) while the cluster administrators are in charge of storage plugins,
+networking plugins and container runtime handlers.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Surface the beginning of pod sandbox creation by Kubelet as a pod condition.
+- Surface the completion of pod sandbox creation by Kubelet as a pod condition.
```

Comment:
What about sandbox teardown?

Reply:
Reporting pod conditions around sandbox teardown is not a goal for this KEP. I will add that as a non-goal below.

Reply:
Addressed [here](https://github.com/kubernetes/enhancements/pull/3087/files#diff-08c9a02aebe0409f9d02d84d2a337e642394319991f684d82798f400510a6d85R270)

Reply:
Graceful sandbox teardown as part of pod deletion can be detected by both of these conditions reporting `false` and pod's `deletionTimestamp` set. Clarified this fact in the KEP as well.

Reply:
I don't understand why this only applies to graceful teardown? What about the non-graceful teardown?